crates/runtime/src/instance/allocator/pooling/uffd.rs

```diff
@@ -0,0 +1,833 @@
+        let tables_end = tables_start + instances.tables.mapping.len();
+        let stacks_start = stacks.mapping.as_ptr() as usize;
+        let stacks_end = stacks_start + stacks.mapping.len();
+        let stack_size = stacks.stack_size;
+
+        // Should always have instances
+        debug_assert!(instances_start != 0);
+
+        Self {
+            instances_start,
+            instance_size: instances.instance_size,
+            max_instances: instances.max_instances,
+            memories_start,
+            memories_end,
+            memory_size: instances.memories.memory_size,
+            max_memories: instances.memories.max_memories,
+            tables_start,
+            tables_end,
+            table_size: instances.tables.table_size,
+            stacks_start,
+            stacks_end,
+            stack_size,
+            page_size: instances.tables.page_size,
+        }
+    }
+
+    /// This is super-duper unsafe as it is used from the handler thread
+    /// to access instance data without any locking primitives.
+    ///
+    /// It is assumed that the thread that owns the instance being accessed is
+    /// currently suspended waiting on a fault to be handled.
+    ///
+    /// Of course a stray faulting memory access from a thread that does not own
+    /// the instance might introduce a race, but this implementation considers
+    /// such to be a serious bug.
+    ///
+    /// If the assumption holds true, accessing the instance data from the handler thread
+    /// should, in theory, be safe.
+    unsafe fn get_instance(&self, index: usize) -> &Instance {
+        debug_assert!(index < self.max_instances);
+        &*((self.instances_start + (index * self.instance_size)) as *const Instance)
+    }
+
+    unsafe fn get_location(&self, addr: usize) -> Option<AddressLocation> {
+        // Check for a memory location
+        if addr >= self.memories_start && addr < self.memories_end {
+            let index = (addr - self.memories_start) / self.memory_size;
+            let memory_index = index % self.max_memories;
+            let memory_start = self.memories_start + (index * self.memory_size);
+            let page_index = (addr - memory_start) / WASM_PAGE_SIZE;
+            let instance = self.get_instance(index / self.max_memories);
+
+            let init_page_index = instance
+                .memories
+                .get(
+                    DefinedMemoryIndex::from_u32(memory_index as u32)
+                        .try_into()
+                        .unwrap(),
+                )
+                .and_then(|m| {
+                    if page_index < m.size() as usize {
+                        Some(page_index)
+                    } else {
+                        None
+                    }
+                });
+
+            return Some(AddressLocation::MemoryPage {
+                page_addr: (memory_start + page_index * WASM_PAGE_SIZE) as _,
+                len: WASM_PAGE_SIZE,
```

Comment:
Sort of similar to above, how come a length of wasm page size is assumed? I would naively assume that the wasm page size wouldn't factor into anything in this module, only perhaps as a multiplier to by default initialize more than one page at a time if one faults.

Reply:
The uffd handler initializes per Wasm page rather than handling a fault per OS page when faulting on linear memories.

See above comment too.